sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,841 @@
+particular place within that buffer and once the constant is requested in
+device code, the compiler generates a load from the same place of the buffer.
+
+Summarizing, overall design looks like:
+
+DPC++ Headers provide special markup, which used by the compiler to detect
+presence of specialization constants and properly handle them.
+
+DPC++ FE handles `kernel_handler` SYCL kernel function argument, creates
+additional kernel argument to pass specialization constants through buffer if
+necessary (i.e. if native support is not available) and generates integration
+footer.
+
+`sycl-post-link` transforms device code to either generate proper SPIR-V
+Friendly IR with specialization constants (when native support is available) or
+to generate correct access to corresponding kernel argument (which are used
+when native support is not available); also the tool generates some device image
+properties with all information needed for DPC++ RT (like which numeric SPIR-V
+ID was assigned to which symbolic ID).
+
+With help of `clang-offload-wrapper` tool, those device image properties are
+embedded into the application together with device code and used by DPC++ RT
+while handling specialization constants during application execution: it either
+calls corresponding PI API to set a value of a specialization constant or it
+fills a special buffer with values of specialization constants and passes it as
+kernel argument to emulate support of specialization constants.
+
+Sections below describe each component in more details.
+
+### DPC++ Headers
+
+DPC++ Headers provide required definitions of `specialization_id` and
+`kernel_handler` classes as well as of many other classes and methods.
+
+`kernel_handler::get_specialization_constant` method, which provides an access
+to specialization constants within device code implements an interface between
+DPC++ Headers and the compiler (`sycl-post-link` tool): it contains a special
+markup, which allows the compiler to detect specialization constants in the
+device code and properly handle them.
+
+```
+namespace sycl {
+// TODO: Add `const` to `DefaultValue` and `RTBuffer`?
+template<typename T>
+T __sycl_getScalar2020SpecConstantValue<T>(const char *SymbolicID, void *DefaultValue, void *RTBuffer);
+template<typename T>
+T __sycl_getComposite2020SpecConstantValue<T>(const char *SymbolicID, void *DefaultValue, void *RTBuffer);
+
+class kernel_handler {
+public:
+  template<auto& SpecName>
+  typename std::remove_reference_t<decltype(SpecName)>::type get_specialization_constant() {
+#ifdef __SYCL_DEVICE_ONLY__
+    return get_on_device<SpecName>();
+#else
+    // some fallback implementation in case this code is launched on host
+#endif __SYCL_DEVICE_ONLY__
+  }
+
+private:
+#ifdef __SYCL_DEVICE_ONLY__
+  template<auto &SpecName, typename T = std::remove_reference_t<decltype(SpecName)>::type>
+  // enable_if T is a scalar type
+  T get_on_device() {
+    auto ID = __builtin_unqiue_ID(SpecName);
```

Comment:
I'm not sure you can call `__builtin_unique_ID()` here.  See my note above about how the string probably needs to be a GUID if the variable has internal linkage.  If the FE generates a GUID in the integration footer, can you be guaranteed to get the same GUID when calling `__builtin_unique_ID()`?

It seems like it would be safer to call the integration footer here like:

```
const char *ID = detail::get_spec_constant_symbolic_ID<SpecName>();
```

Reply:
> I'm not sure you can call `__builtin_unique_ID()` here. See my note above about how the string probably needs to be a GUID if the variable has internal linkage.

Why not? The whole point of the compiler built-in is that it is capable to return the right string, accounting for the variable linkage.

> If the FE generates a GUID in the integration footer, can you be guaranteed to get the same GUID when calling `__builtin_unique_ID()`?

This is assumption I'm making. For example for existing `__builtin_unique_stable_name` FE is capable to emit the _result_ of call to this built-in without having the actual call: we do that in our regular integration header.

> It seems like it would be safer to call the integration footer here like:

I think we need some input from @erichkeane before going this way: Erich, would it be possible to generate the header while we compile the source file, then attach generated piece of code at the end on the fly and proceed with compilation? If that is possible, then yes, we can re-use integration footer here

Reply:
I don't think we can put it at the end of the device compile 'on the fly', by the time we generate an integration header, our lexer is 'done'.  The purpose of the integration footer is that it is the 'footer' is available at the end of the host compile, right?

Reply:
>  The purpose of the integration footer is that it is the 'footer' is available at the end of the host compile, right?

Right, the original proposal assumed that it will be only used for host compilation

Reply:
I assume that we have reached agreement on this topic, so I will mark this discussion as resolved